### PR TITLE
Adjust chpl to build with compiler/next library

### PR DESCRIPTION
**Pull request by mppf**
This PR adjusts `chpl` to build with the compiler/next compiler library.

The compiler/next directory is managed by cmake (requiring 3.14.3) but at
this time I don't think we are ready to add a requirement that people
building Chapel have cmake. Somebody building with `CHPL_LLVM=bundled`
will need to have cmake (and 3.14.3 as of LLVM 12) but we also have
quickstart configurations and will encourage people to use
`CHPL_LLVM=system` anyway. So, for now, this PR takes the approach of
creating a parallel Makefiles system for compiler/next. The Makefiles
added in this PR serve only the purpose of including the compiler/next
code in the `chpl` program. They don't build an actual .a or .so library,
they don't generate the `config.h` with a version number, they don't
build documentation, they can't regenerate the parser, and they don't
build or run the C++ compiler tests. Those activities will still be done
by the `cmake` version.

This PR adds a developer-only flag `--compiler-library-parser` to use the
compiler/next parser. We can use this flag to start checking to see which
test files the compiler/next parser handles. In the future it will also
enable translation from uAST to the production AST which is a next
milestone.

Since the compiler/next source code requires C++14, this PR adjusts
make/compiler/Makefile.gnu and make/compiler/Makefile.clang to include
the appropriate argument to request C++14. C++14 support is already
documented as a requirement as of PR #17604.

Since the compiler/next source code includes a C source file (actually
just one), this PR adjusts make/compiler/Makefile.gnu and
make/compiler/Makefile.clang to set `COMP_CFLAGS` for use when compiling
that file.

In order to link both the production parser and the compiler/next parser
into one program, we needed to change the names of symbols and headers.
So, this PR renames these files (as well as chapel.lex/chapel.ypp to
avoid duplicate file names in different directories) and uses flex/bison
options to specify a prefix `yychpl_` instead of `yy` so that symbol
names are different to avoid link errors.

Since the compiler/next parser now needs to exist in order for `chpl` to
build, this PR stops `.gitignoring` the generated files and commits an
initial version of them. Additionally, this PR uses a bison option to
require bison 3.7 in order to reduce the number of changes to the
generated parser (the changes are big when bison versions change) and to
reduce the amount of data it adds to the git history.

Lastly, this PR fixes two bugs:
* Fix unused nonterminal from PR #17815 which lead to bison warnings
* Fix problem in compiler/next error reporting where all errors were
  print starting with two spaces that was introduced in PR #17648

Reviewed by @dlongnecke-cray - thanks!

- [x] full local testing

### CODE REVIEW

File: compiler/next/lib/uast/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+NEXT_UAST_OBJDIR = $(COMPILER_BUILD)/next/lib/uast
+
+ALL_SRCS += next/lib/uast/*.cpp
+
+NEXT_UAST_SRCS =                                 \
+  ASTList.cpp \
```
Comment (dlongnecke-cray) [inline compiler/next/lib/uast/Makefile.include:24]: 😭

File: compiler/parser/parser.cpp [hunk 2]
```diff
@@ -751,6 +764,33 @@ static ModuleSymbol* parseFile(const char* path,
   return retval;
 }
 
+
+static ModuleSymbol* uASTParseFile(const char* fileName,
```
Comment (dlongnecke-cray) [inline compiler/parser/parser.cpp:768]: :tada::tada::tada::tada:

**Review by dlongnecke-cray — APPROVED** [review]
I've gotta get my two loop PRs in before you commit this so I don't have 5000 line diffs ;d

**Comment by dlongnecke-cray** [comment]
Does rebuilding the parser only cause a lot of churn when switching minor versions? Will it be relatively "fine" if I'm using `3.7.5`?

**Comment by mppf** [comment]
> Does rebuilding the parser only cause a lot of churn when switching minor versions? Will it be relatively "fine" if I'm using 3.7.5?

I do not know but I expect we will learn soon enough

**Pull request merged**